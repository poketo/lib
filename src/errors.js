// @flow

type ErrorCode =
  | 'INVALID_URL'
  | 'UNSUPPORTED_SITE'
  | 'UNSUPPORTED_SITE_REQUEST'
  | 'TIMEOUT'
  | 'HTTP_ERROR';

class PoketoError extends Error {
  code: ErrorCode;
  constructor(errorCode: ErrorCode, message: string) {
    super(message);
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.code = errorCode;
  }
}

class InvalidUrlError extends PoketoError {
  constructor(url: string) {
    super('INVALID_URL', `Unable to parse url '${url}'`);
  }
}

class HTTPError extends PoketoError {
  statusCode: number;
  url: string;

  constructor(statusCode: number, message: string, url: string) {
    super('HTTP_ERROR', message);
    this.statusCode = statusCode;
    this.url = url;
  }
}

class NotFoundError extends HTTPError {
  constructor(url: string) {
    super(404, 'Not Found', url);
  }
}

class TimeoutError extends PoketoError {
  constructor(message: string, url: string) {
    super('TIMEOUT', message);
  }
}

class UnsupportedSiteError extends PoketoError {
  constructor(site: string) {
    super('UNSUPPORTED_SITE', `Site at '${site}' is not supported`);
  }
}

class UnsupportedSiteRequestError extends PoketoError {
  constructor(siteName: string, operationName: string) {
    super(
      'UNSUPPORTED_SITE_REQUEST',
      `${siteName} does not support ${operationName}`,
    );
  }
}

export default {
  PoketoError,
  HTTPError,
  InvalidUrlError,
  NotFoundError,
  TimeoutError,
  UnsupportedSiteError,
  UnsupportedSiteRequestError,
};
