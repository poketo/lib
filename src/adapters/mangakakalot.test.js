import site from './mangakakalot';
import errors from '../errors';

describe('MangakakalotAdapter', () => {
  const server = new AdapterVcrServer(site);

  beforeAll(async () => {
    await server.listenAndMock(57171);
  });

  afterAll(() => {
    server.close();
  });

  describe('parseUrl', () => {
    it('returns the components of a url', () => {
      expect(
        site.parseUrl('http://mangakakalot.com/manga/urami_koi_koi_urami_koi/'),
      ).toEqual({ seriesSlug: 'urami_koi_koi_urami_koi', chapterSlug: null });

      expect(
        site.parseUrl(
          'http://mangakakalot.com/chapter/urami_koi_koi_urami_koi/chapter_15.2',
        ),
      ).toEqual({
        seriesSlug: 'urami_koi_koi_urami_koi',
        chapterSlug: 'chapter_15.2',
      });
    });

    it('throws on unparseable urls', () => {
      expect(() => {
        site.parseUrl('https://www.mangaupdates.com/authors.html?id=462621');
      }).toThrow(errors.InvalidUrlError);
    });
  });

  describe('getSeries', () => {
    it('returns a metadata object', async () => {
      const { chapters, ...metadata } = await site.getSeries(
        'urami_koi_koi_urami_koi',
      );

      expect(metadata).toMatchSnapshot({
        updatedAt: expect.any(Number),
      });

      const chapterNumbersToTest = ['38'];
      const chaptersToTest = chapters.filter(chapter =>
        chapterNumbersToTest.includes(chapter.chapterNumber),
      );
      expect(chaptersToTest).toMatchSnapshot();
    });
  });

  describe('getChapter', () => {
    it('returns a list of pages', async () => {
      const chapter = await site.getChapter(
        'urami_koi_koi_urami_koi',
        'chapter_5',
      );

      expect(chapter.pages).toHaveLength(48);
      expect(chapter.pages).toEqual(
        expect.arrayContaining([
          {
            id: expect.any(String),
            url: expect.stringContaining('http'),
          },
        ]),
      );
    });
  });
});
