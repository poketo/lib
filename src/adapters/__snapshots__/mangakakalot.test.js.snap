// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MangakakalotAdapter getSeries returns a metadata object 1`] = `
Object {
  "coverImageUrl": "http://avatar.mkklcdnv2.com/avatar_manga/3566-urami_koi_koi_urami_koi.jpg",
  "slug": "urami_koi_koi_urami_koi",
  "title": "Urami Koi, Koi, Urami Koi.",
  "updatedAt": Any<Number>,
  "url": "http://localhost:57171/manga/urami_koi_koi_urami_koi",
}
`;

exports[`MangakakalotAdapter getSeries returns a metadata object 2`] = `
Array [
  Object {
    "chapterNumber": "38",
    "createdAt": 1514196540,
    "slug": "chapter_38",
    "title": undefined,
    "url": "http://localhost:57171/chapter/urami_koi_koi_urami_koi/chapter_38",
  },
]
`;
